File: Tracks/csharp/perfect-numbers/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect = 0,
    Abundant = 1,
    Deficient = -1
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0)
        {
            throw new ArgumentOutOfRangeException();
        }

        if (number == 1)
        {
            return Classification.Deficient;
        }

        return (Classification)Enumerable.Range(1, number / 2)
                                         .Where(current => number % current == 0)
                                         .Sum()
                                         .CompareTo(number);
    }
}

File: Tracks/typescript/perfect-numbers/perfect-numbers.ts
type NumberType = "perfect" | "abundant" | "deficient";
type Sign = 1 | -1 | 0;
const typeMap = new Map<Sign, NumberType>([
    [-1, "abundant"],
    [1, "deficient"],
    [0, "perfect"]
]);

export default class PerfectNumbers {
    static classify = (candidate: number): NumberType => {
        if (candidate <= 0) {
            throw new Error("Classification is only possible for natural numbers.");
        }
        return typeMap.get(Math.sign(PerfectNumbers.substractAllFactors(candidate)) as Sign) as NumberType;
    };

    private static readonly substractAllFactors = (candidate: number) => {
        let sum = candidate;
        for (let index = 1; index <= candidate / 2; index++) {
            sum -= (candidate % index === 0 && index) || 0;
        }
        return sum;
    };
}
